flow/display_list.h

```diff
@@ -0,0 +1,427 @@
+  V(ClipRRect)                      \
+  V(ClipPath)                       \
+                                    \
+  V(DrawPaint)                      \
+  V(DrawColor)                      \
+                                    \
+  V(DrawLine)                       \
+  V(DrawRect)                       \
+  V(DrawOval)                       \
+  V(DrawCircle)                     \
+  V(DrawRRect)                      \
+  V(DrawDRRect)                     \
+  V(DrawArc)                        \
+  V(DrawPath)                       \
+                                    \
+  V(DrawPoints)                     \
+  V(DrawLines)                      \
+  V(DrawPolygon)                    \
+  V(DrawVertices)                   \
+                                    \
+  V(DrawImage)                      \
+  V(DrawImageRect)                  \
+  V(DrawImageNine)                  \
+  V(DrawImageLattice)               \
+  V(DrawAtlas)                      \
+  V(DrawAtlasColored)               \
+  V(DrawAtlasCulled)                \
+  V(DrawAtlasColoredCulled)         \
+                                    \
+  V(DrawSkPicture)                  \
+  V(DrawDisplayList)                \
+  V(DrawTextBlob)                   \
+  /* V(DrawShadowRec) */            \
+                                    \
+  V(DrawShadow)
+
+#define DL_OP_TO_ENUM_VALUE(name) name,
+enum DisplayListOpType { FOR_EACH_DISPLAY_LIST_OP(DL_OP_TO_ENUM_VALUE) };
+#undef DL_OP_TO_ENUM_VALUE
+
+class Dispatcher;
+class DisplayListBuilder;
+
+// The base class that contains a sequence of rendering operations
+// for dispatch to a Dispatcher. These objects must be instantiated
+// through an instance of DisplayListBuilder::build().
+class DisplayList : public SkRefCnt {
+ public:
+  static const SkSamplingOptions NearestSampling;
+  static const SkSamplingOptions LinearSampling;
+  static const SkSamplingOptions MipmapSampling;
+  static const SkSamplingOptions CubicSampling;
+
+  DisplayList()
+      : ptr_(nullptr),
+        used_(0),
+        opCount_(0),
+        uniqueID_(0),
+        bounds_({0, 0, 0, 0}),
+        boundsCull_({0, 0, 0, 0}) {}
+
+  ~DisplayList();
+
+  void dispatch(Dispatcher& ctx) const { dispatch(ctx, ptr_, ptr_ + used_); }
+
+  void renderTo(SkCanvas* canvas) const;
+
+  size_t bytes() const { return used_; }
+  int opCount() const { return opCount_; }
+  uint32_t uniqueID() const { return uniqueID_; }
+
+  const SkRect& bounds() {
+    if (bounds_.width() < 0.0) {
+      // computeBounds() will leave the variable with a
+      // non-negative width and height
+      computeBounds();
+    }
+    return bounds_;
+  }
+
+  bool equals(const DisplayList& other) const;
+
+ private:
+  DisplayList(uint8_t* ptr, size_t used, int opCount, const SkRect& cullRect);
+
+  uint8_t* ptr_;
+  size_t used_;
+  int opCount_;
+
+  uint32_t uniqueID_;
+  SkRect bounds_;
+
+  // Only used for drawPaint() and drawColor()
+  SkRect boundsCull_;
```

Comment:
```suggestion
  SkRect bounds_cull_;
```

Here and elsewhere: use snake_case for variable names.